File: backend/ColorGenerator.py
from tkinter import *
import math

# updated by Jun
# color scheme still needs discussion

def findAngle(x,y): 
    radian = math.atan2(y,x)
    degree = radian * (360/(2*math.pi))
    if (degree < 0): 
        degree = 360 + degree
    return degree


def findColor(degree): 
    red = 120 
    blue = 240
    green = 360
    rDif = red-degree
    rValue = abs(round(255-rDif*2.125))
    bDif = blue-degree
    bValue = abs(round(255-bDif*2.125))
    gDif = green-degree
    gValue = abs(round(255-gDif*2.125))
    values = (rValue,gValue,bValue)
    finalValues = []
    for value in values: 
        if (value > 255): 
            
            value = 0 
        
        finalValues += [value]

    return finalValues


def rgbString(red,green,blue): 
        return "#%02x%02x%02x" % (red,green,blue)


def draw(canvas, i, color):
    size = 100
    lo = 10 + i*size
    hi = 10 + (i+1)*size
    canvas.create_rectangle(lo, lo, hi, hi,fill=color)

def runDrawing(width, height,colors):
    root = Tk()
    canvas = Canvas(root, width=width, height=height)
    canvas.pack()
    for i in range(len(colors)):
        draw(canvas, i, colors[i])
    root.mainloop()


def getRGB(x,y): 
    angle = findAngle(x,y) 
    color = findColor(angle) 
    red = color[0]
    green = color[1]
    blue = color[2] 
    rgb = rgbString(red,green,blue)
    return rgb
    
def run():
    points = [(1,1), (1,-1), (-1,1), (-1,-1)]
    colors = []
    for point in points:
        (x, y) = point
        rgb = getRGB(x, y)
        colors.append(rgb)
    runDrawing(600, 600,colors)

File: backend/ColorGenerator.js

/* (x, y)
 * (+, +): happy | high energy
 * (+, -): happy | low energy
 * (-, +): sad | high energy
 * (-, -): sad | low energy
*/


function findAngle (x, y) {
  var rad = Math.atan2(y, x);
  if (rad < 0) {
    rad += Math.PI * 2; 
  }
  return {radian: rad} 
}

function findColor (x, y, diag) {
  var intensity = Math.hypot(x, y);
  var radian = findAngle(x, y);
  rad = radian.radian; 
  
  // + +
  if (rad >= 0 && rad < Math.PI/2 ) {
    console.log("green");
    // green
    r = 255 * (1 - intensity/diag);
    b = 255 * (1 - intensity/diag);
    g = 255;
    if (rad >= Math.PI/4) {
      r += (255/4) * (rad-Math.PI/4)/(Math.PI/4);
      g -= (255/4) * (rad-Math.PI/4)/(Math.PI/4);
      
    } 
    else {
      r += (255/4) * (1 - rad/(Math.PI/4));
      b += (255/4) * (1 - rad/(Math.PI/4));
      g -= (255/4) * (1 - rad/(Math.PI/4));
    }
  }
  
  // - +
  else if (rad >= Math.PI/2 && rad < Math.PI ) {
    // red
    console.log("red");
    r = 255;
    g = 255 * (1 - intensity/diag);
    b = 255 * (1 - intensity/diag);
    if (rad <= Math.PI * 3/4) {
      r -= (255/2) * (1 - (rad-Math.PI/2)/(Math.PI/4));
      g += (255/2) * (1 - (rad-Math.PI/2)/(Math.PI/4));
      
    } 
    else {
      r -= (255/2) * (1 - (Math.PI - rad)/(Math.PI/4));
      b += (255/2) * (1 - (Math.PI - rad)/(Math.PI/4));
    }
  }

  // - -
  else if (rad >= Math.PI && rad < Math.PI * 3/2 ) {
    // blue
    console.log("blue");
    r = 255 * (1 - intensity/diag);
    g = 255 * (1 - intensity/diag);
    b = 255;
    if (rad <= Math.PI * 5/4) {
      r += (255/2) * (1 - (rad - Math.PI))/(Math.PI/4);
      b -= (255/2) * (1 - (rad - Math.PI))/(Math.PI/4);
      
    } 
    else {
      r += (255/2) * (rad - Math.PI * 5/4)/(Math.PI/4);
    }
  }
  
  // + + 
  else {
    console.log("pink");
    r = 255;
    g = 255 * (1 - intensity/diag);
    b = 255;
    if (rad >= Math.PI * 7/4) {
      r -= (255/2) * (rad-Math.PI * 7/4)/(Math.PI/4);
      b -= (255/2) * (rad-Math.PI * 7/4)/(Math.PI/4);
      g += (255/2) * (rad-Math.PI * 7/4)/(Math.PI/4);
    } 
    else {
      r -= (255/2) * (1 - (rad - Math.PI * 3/2)/(Math.PI/4));
    }
  }

  r = Math.floor(Math.min(255, r));
  g = Math.floor(Math.min(255, g));
  b = Math.floor(Math.min(255, b));

  return {r : r, g : g, b : b};
}
